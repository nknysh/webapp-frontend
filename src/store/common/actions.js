import { curry, pipe, propOr, objOf, pathOr, forEach } from 'ramda';

import { APP_VERBOSE_ERRORS } from 'config';
import { isArray } from 'utils';

import { getErrorActionName, getIdleActionName, getSuccessActionName, getLoadingActionName } from 'store/utils';
import { enqueueNotification } from 'store/modules/ui/actions';

export const STATUS_TO_IDLE = 'STATUS_TO_IDLE';
export const STORE_RESET = 'STORE_RESET';
export const PAGE_CHANGE = 'PAGE_CHANGE';

/**
 * Reset statuses action
 *
 * @returns {object}
 */
export const resetStatuses = () => ({
  type: STATUS_TO_IDLE,
});

/**
 * Store reset action
 *
 * @returns {object}
 */
export const storeReset = () => ({
  type: STORE_RESET,
});

/**
 * Page change action
 *
 * @param {*} payload
 */
export const pageChange = payload => ({
  type: PAGE_CHANGE,
  payload,
});

/**
 * Idle action
 *
 * Curried function that creates an idle action object
 * based on the action type passed in
 *
 * @param {string} type The action type e.g. `'STORE_RESET'`
 * @param {array | object} data the payload
 * @returns {Function | object}
 */
export const idleAction = curry((type, data) => ({
  type: getIdleActionName(type),
  payload: isArray(data) ? [...data] : { ...data },
}));

/**
 * Loading action
 *
 * Curried function that creates an loading action object
 * based on the action type passed in
 *
 * @param {string} type The action type e.g. `'STORE_RESET'`
 * @param {array | object} data the payload
 * @returns {Function | object}
 */
export const loadingAction = curry((type, data) => ({
  type: getLoadingActionName(type),
  payload: isArray(data) ? [...data] : { ...data },
}));

/**
 * Success action
 *
 * Curried function that creates an success action object
 * based on the action type passed in
 *
 * @param {string} type The action type e.g. `'STORE_RESET'`
 * @param {array | object} data the payload
 * @returns {Function | object}
 */
export const successAction = curry((type, data) => ({
  type: getSuccessActionName(type),
  payload: isArray(data) ? [...data] : { ...data },
}));

/**
 * Error action
 *
 * Curried function that creates an error action object
 * based on the action type passed in
 *
 * @param {string} type The action type e.g. `'STORE_RESET'`
 * @param {array | object} data the payload
 * @returns {Function | object}
 */
export const errorAction = curry((type, error) => ({
  type: getErrorActionName(type),
  payload: isArray(error) ? [...error] : { ...error },
}));

/**
 * Entities object
 *
 * Curried function that creates an object like `{ entities: { [type]: data }}`
 *
 * @param {string} type Final key
 * @param {*} data
 * @returns {Function | object}
 */
export const entitiesObject = curry((type, data) =>
  pipe(
    objOf(type),
    objOf('entities')
  )(data)
);

/**
 * Error from response action
 *
 * Thunk that will dispatch notifications and deal
 * with errors from API call responses
 *
 * @param {object} action
 * @param {object} error
 * @param {string} defaultMessage
 */
export const errorFromResponse = (action, error, defaultMessage) => async dispatch => {
  dispatch(errorAction(action, propOr({ message: propOr('Unknown error', 'message', error) }, 'response', error)));

  // Extracts the errors from the error response
  const errors = pathOr([], ['response', 'data', 'errors'], error);

  // If there was a default message to show, then enqueue a notification
  // that will show to the user
  defaultMessage &&
    dispatch(
      enqueueNotification({
        message: defaultMessage,
        options: { variant: 'error' },
      })
    );

  // If the app should show verbose errors then run through all the errors
  // from the response and show them
  APP_VERBOSE_ERRORS &&
    forEach(
      ({ detail, message }) =>
        dispatch(
          enqueueNotification({
            message: message || detail || 'Unknown Error',
            options: { variant: 'error' },
          })
        ),
      errors
    );
};

/**
 * Generic action
 *
 * @param {string} type
 * @param {*} payload
 * @returns {object}
 */
export const genericAction = (type, payload) => ({
  type,
  payload,
});
