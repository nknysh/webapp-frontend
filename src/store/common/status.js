import { equals, anyPass } from 'ramda';

export const Status = Object.freeze({
  LOADING: 'LOADING',
  IDLE: 'IDLE',
  ERROR: 'ERROR',
  SUCCESS: 'SUCCESS',
  SENDING: 'SENDING',
});

export const isIdle = equals(Status.IDLE);
export const isLoading = equals(Status.LOADING);
export const isError = equals(Status.ERROR);
export const isSuccess = equals(Status.SUCCESS);
export const isSending = equals(Status.SENDING);

export const isActive = anyPass([isLoading, isSending]);
