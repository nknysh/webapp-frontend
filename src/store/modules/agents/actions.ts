import { ICompany, ITravelAgent } from 'services/BackendApi';
export const GET_TRAVEL_AGENTS_REQUEST = 'agentsModule/GET_TRAVEL_AGENTS_REQUEST';
export const GET_TRAVEL_AGENTS_SUCCESS = 'agentsModule/GET_TRAVEL_AGENTS_SUCCESS';
export const GET_TRAVEL_AGENTS_FAILURE = 'agentsModule/GET_TRAVEL_AGENTS_FAILURE';
export const SELECTED_TA_CHANGE = 'agentsModule/SELECTED_TA_CHANGE';
export const SHOW_TA_DROPDOWN = 'agentsModule/SHOW_TA_DROPDOWN';
export const SEARCH_TA_BY_NAME_CHANGE = 'agentsModule/SEARCH_TA_BY_NAME_CHANGE';
export const GET_COMPANIES_REQUEST = 'agentsModule/GET_COMPANIES_REQUEST';
export const GET_COMPANIES_SUCCESS = 'agentsModule/GET_COMPANIES_SUCCESS';
export const GET_COMPANIES_FAILURE = 'agentsModule/GET_COMPANIES_FAILURE';
export const SELECTED_COMPANY_CHANGE = 'agentsModule/SELECTED_COMPANY_CHANGE';
export const SHOW_COMPANY_DROPDOWN = 'agentsModule/SHOW_COMPANY_DROPDOWN';
export const SEARCH_COMPANY_BY_NAME = 'agentsModule/SEARCH_COMPANY_BY_NAME';

export type GetTravelAgentRequestAction = ReturnType<typeof getTravelAgentsRequestAction>;
export const getTravelAgentsRequestAction = (forceLoad?: boolean) => ({
  type: GET_TRAVEL_AGENTS_REQUEST as typeof GET_TRAVEL_AGENTS_REQUEST,
  forceLoad,
});

export type GetTravelAgentSuccessAction = ReturnType<typeof getTravelAgentsSuccessAction>;
export const getTravelAgentsSuccessAction = (agents: ITravelAgent[]) => ({
  type: GET_TRAVEL_AGENTS_SUCCESS as typeof GET_TRAVEL_AGENTS_SUCCESS,
  agents,
});

export type GetTravelAgentFailureAction = ReturnType<typeof getTravelAgentsFailureAction>;
export const getTravelAgentsFailureAction = (error: string) => ({
  type: GET_TRAVEL_AGENTS_FAILURE as typeof GET_TRAVEL_AGENTS_FAILURE,
  error,
});

export type SelectedTaChangeAction = ReturnType<typeof selectedTaChangeAction>;
export const selectedTaChangeAction = (value: ITravelAgent | null) => ({
  type: SELECTED_TA_CHANGE as typeof SELECTED_TA_CHANGE,
  value,
});

export type ShowTaDropdownAction = ReturnType<typeof showTaDropdownAction>;
export const showTaDropdownAction = (value: boolean) => ({
  type: SHOW_TA_DROPDOWN as typeof SHOW_TA_DROPDOWN,
  value,
});

export type SearchTaByNameChangeAction = ReturnType<typeof searchTaByNameChangeAction>;
export const searchTaByNameChangeAction = (value: string) => ({
  type: SEARCH_TA_BY_NAME_CHANGE as typeof SEARCH_TA_BY_NAME_CHANGE,
  value,
});


export type GetCompaniesRequestAction = ReturnType<typeof getCompaniesRequestAction>;
export const getCompaniesRequestAction = () => ({
  type: GET_COMPANIES_REQUEST as typeof GET_COMPANIES_REQUEST,
});

export type GetCompaniesSuccessAction = ReturnType<typeof getCompaniesSuccessAction>;
export const getCompaniesSuccessAction = (companies: ICompany[]) => ({
  type: GET_COMPANIES_SUCCESS as typeof GET_COMPANIES_SUCCESS,
  companies,
});

export type GetCompaniesFailureAction = ReturnType<typeof getCompaniesFailureAction>;
export const getCompaniesFailureAction = (error: string) => ({
  type: GET_COMPANIES_FAILURE as typeof GET_COMPANIES_FAILURE,
  error,
});

export type SelectedCompanyChangeAction = ReturnType<typeof selectedCompanyChangeAction>;
export const selectedCompanyChangeAction = (value: ICompany) => ({
  type: SELECTED_COMPANY_CHANGE as typeof SELECTED_COMPANY_CHANGE,
  value,
});

export type ShowCompanyDropdownAction = ReturnType<typeof showCompanyDropdownAction>;
export const showCompanyDropdownAction = (value: boolean) => ({
  type: SHOW_COMPANY_DROPDOWN as typeof SHOW_COMPANY_DROPDOWN,
  value,
});

export type SearchCompanyByNameAction = ReturnType<typeof searchCompanyByNameAction>;
export const searchCompanyByNameAction = (value: string) => ({
  type: SEARCH_COMPANY_BY_NAME as typeof SEARCH_COMPANY_BY_NAME,
  value,
});

export type AgentsModuleAction =
  | GetTravelAgentRequestAction
  | GetTravelAgentSuccessAction
  | GetTravelAgentFailureAction
  | SelectedTaChangeAction
  | ShowTaDropdownAction
  | SearchTaByNameChangeAction
  | GetCompaniesRequestAction
  | GetCompaniesSuccessAction
  | GetCompaniesFailureAction
  | SelectedCompanyChangeAction
  | ShowCompanyDropdownAction
  | SearchCompanyByNameAction;
