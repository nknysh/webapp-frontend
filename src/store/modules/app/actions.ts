export const SET_LATEST_APP_VERSION = 'app/SET_LATEST_APP_VERSION';

export type SetLatestAppVersion = ReturnType<typeof setLatestAppVersion>;
export const setLatestAppVersion = (latestAppVersion: string, isAppDeprecated: boolean) => ({
  type: SET_LATEST_APP_VERSION as typeof SET_LATEST_APP_VERSION,
  latestAppVersion,
  isAppDeprecated
});


export type AppAction =
 | SetLatestAppVersion
