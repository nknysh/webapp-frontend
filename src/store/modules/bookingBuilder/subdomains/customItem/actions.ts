export const SHOW_CUSTOM_ITEM_FORM = 'bookingBuilder/customItem/SHOW_CUSTOM_ITEM_FORM';
export const HIDE_CUSTOM_ITEM_FORM = 'bookingBuilder/customItem/HIDE_CUSTOM_ITEM_FORM';

export const UPDATE_CUSTOM_ITEM_NAME = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_NAME';
export const UPDATE_CUSTOM_ITEM_TOTAL = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_TOTAL';
export const UPDATE_CUSTOM_ITEM_DESCRIPTION = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_DESCRIPTION';
export const UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN';
export const UPDATE_CUSTOM_ITEM_COUNTS_AS_TRANSFER = 'bookingBuilder/customItem/UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN';


export type ShowCustomItemFormAction = ReturnType<typeof showCustomItemFormAction>;
export const showCustomItemFormAction = () => ({
  type: SHOW_CUSTOM_ITEM_FORM as typeof SHOW_CUSTOM_ITEM_FORM
});

export type HideCustomItemFormAction = ReturnType<typeof hideCustomItemFormAction>;
export const hideCustomItemFormAction = () => ({
  type: HIDE_CUSTOM_ITEM_FORM as typeof HIDE_CUSTOM_ITEM_FORM
});

//------------------- edit ----------------------------------

export type UpdateCustomItemNameAction = ReturnType<typeof updateCustomItemNameAction>;
export const updateCustomItemNameAction = (value: string) => ({
  type: UPDATE_CUSTOM_ITEM_NAME as typeof UPDATE_CUSTOM_ITEM_NAME,
  value
});

export type UpdateCustomItemTotalAction = ReturnType<typeof updateCustomItemTotalAction>;
export const updateCustomItemTotalAction = (value: string) => ({
  type: UPDATE_CUSTOM_ITEM_TOTAL as typeof UPDATE_CUSTOM_ITEM_TOTAL,
  value
});

export type UpdateCustomItemDescriptionAction = ReturnType<typeof updateCustomItemDescriptionAction>;
export const updateCustomItemDescriptionAction = (value: string) => ({
  type: UPDATE_CUSTOM_ITEM_DESCRIPTION as typeof UPDATE_CUSTOM_ITEM_DESCRIPTION,
  value
});

export type UpdateCustomItemCountsAsMealPlanAction = ReturnType<typeof updateCustomItemCountsAsMealPlanAction>;
export const updateCustomItemCountsAsMealPlanAction = (value: boolean) => ({
  type: UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN as typeof UPDATE_CUSTOM_ITEM_COUNTS_AS_MEAL_PLAN,
  value
});

export type UpdateCustomItemCountsAsTransferAction = ReturnType<typeof updateCustomItemCountsAsTransferAction>;
export const updateCustomItemCountsAsTransferAction = (value: boolean) => ({
  type: UPDATE_CUSTOM_ITEM_COUNTS_AS_TRANSFER as typeof UPDATE_CUSTOM_ITEM_COUNTS_AS_TRANSFER,
  value
});


export type CustomItemAction =
  | ShowCustomItemFormAction
  | HideCustomItemFormAction

  | UpdateCustomItemNameAction
  | UpdateCustomItemTotalAction
  | UpdateCustomItemDescriptionAction
  | UpdateCustomItemCountsAsMealPlanAction
  | UpdateCustomItemCountsAsTransferAction;
