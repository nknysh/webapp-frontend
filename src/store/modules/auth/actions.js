import { prop, propEq, omit, propOr, mergeDeepLeft, equals } from 'ramda';
import { FORBIDDEN } from 'http-status';

import client from 'api/auth';
import userClient from 'api/users';
import { UserStatusTypes, AuthTypes } from 'config/enums';

import { genericAction, successAction, errorAction, errorFromResponse, storeReset } from 'store/common';
import { enqueueNotification } from 'store/modules/ui/actions';
import { getCurrentUserType } from './selectors';

export const AUTH_REQUEST = 'AUTH_REQUEST';
export const AUTH_CHECK = 'AUTH_CHECK';
export const AUTH_OK = 'AUTH_OK';
export const AUTH_RESET = 'AUTH_RESET';
export const AUTH_SET_TOKEN = 'AUTH_SET_TOKEN';
export const AUTH_SIGN_UP = 'AUTH_SIGN_UP';
export const AUTH_LOG_OUT = 'AUTH_LOG_OUT';
export const AUTH_PASSWORD_RESET = 'AUTH_PASSWORD_RESET';
export const AUTH_SET_PASSWORD = 'AUTH_SET_PASSWORD';
export const AUTH_COUNTRY_SET = 'AUTH_COUNTRY_SET';
export const AUTH_PASSWORD_UPDATE = 'AUTH_PASSWORD_UPDATE';

// Localstorage constants for auth
export const AUTH_TOKEN = 'authToken';
export const AUTH_USER = 'authUser';
export const AUTH_COUNTRY_CODE = 'authCountryCode';

/**
 * Auth reset action
 *
 * @returns {object}
 */
const authReset = () => ({
  type: AUTH_RESET,
});

/**
 * Set token action
 *
 * @param {string} token
 * @returns {object}
 */
export const setToken = token => ({
  type: AUTH_SET_TOKEN,
  payload: { token },
});

/**
 * Auth request action
 *
 * @param {*} value
 * @returns {object}
 */
export const authRequest = value => ({
  type: AUTH_REQUEST,
  payload: value,
});

/**
 * Auth password reset action
 *
 * @param {*} value
 * @returns {object}
 */
export const authPasswordReset = value => ({
  type: AUTH_PASSWORD_RESET,
  payload: value,
});

/**
 * Auth set password reset action
 *
 * @param {*} value
 * @returns {object}
 */
export const authSetPasswordReset = value => ({
  type: AUTH_SET_PASSWORD,
  payload: value,
});

/**
 * Auth ok action
 *
 * @param {*} values
 * @returns {object}
 */
export const authOk = values => ({
  type: AUTH_OK,
  payload: values,
});

/**
 * Auth sign up action
 *
 * @param {*} values
 * @returns {object}
 */
export const authSignUp = values => ({
  type: AUTH_SIGN_UP,
  payload: values,
});

/**
 * Auth log out action
 *
 * @param {string} token
 * @returns {object}
 */
export const authLogOut = token => ({
  type: AUTH_LOG_OUT,
  payload: token,
});

/**
 * Auth set country action
 *
 * @param {string} countryCode
 * @returns {object}
 */
export const authSetCountryAction = countryCode => ({
  type: AUTH_COUNTRY_SET,
  payload: countryCode,
});

/**
 * Set remembered token
 *
 * @param {string} token
 */
export const setRememberedToken = token => localStorage.setItem(AUTH_TOKEN, token);

/**
 * Set remembered user
 *
 * @param {object} user
 */
export const setRememberedUser = user => localStorage.setItem(AUTH_USER, JSON.stringify(user));

/**
 * Set remembered country
 *
 * @param {string} countryCode
 */
export const setRememberedCountry = countryCode => localStorage.setItem(AUTH_COUNTRY_CODE, countryCode);

/**
 * Delete remembered token
 */
export const deleteRememberedToken = () => localStorage.removeItem(AUTH_TOKEN);

/**
 * Delete remembered user
 */
export const deleteRememberedUser = () => localStorage.removeItem(AUTH_USER);

/**
 * Delete remembered country
 */
export const deleteRememberedCountry = () => localStorage.removeItem(AUTH_COUNTRY_CODE);

/**
 * Auth set country action
 *
 * Sets the remembered country into local storage
 *
 * @param {string} countryCode
 * @returns {Function}
 */
export const authSetCountry = countryCode => dispatch => {
  setRememberedCountry(countryCode);
  dispatch(authSetCountryAction(countryCode));
};

/**
 * Get user action
 *
 * @param {Function} dispatch
 */
export const getUser = async dispatch => {
  dispatch(authRequest());

  try {
    const { data } = await client.getUser();
    dispatch(successAction(AUTH_REQUEST, data));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_REQUEST, e, 'There was a problem fetching the current user.'));
  }
};

/**
 * Sign up action
 *
 * @param {object} values
 * @returns {Function}
 */
export const signUp = values => async dispatch => {
  dispatch(authSignUp(values));

  try {
    const { data } = await client.signUp({ data: { attributes: values } });
    dispatch(successAction(AUTH_SIGN_UP, data));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SIGN_UP, e));
  }
};

/**
 * Clear user action
 *
 * Removes user form local storage
 *
 * @param {Function} dispatch
 */
const clearUser = dispatch => {
  deleteRememberedToken();
  deleteRememberedUser();
  deleteRememberedCountry();

  dispatch(authReset());
  dispatch(successAction(AUTH_LOG_OUT, {}));
  dispatch(storeReset());
};

/**
 * Log out action
 *
 * Trrigger logout request
 *
 * @param {string} token
 * @returns {Function}
 */
export const logOut = token => async dispatch => {
  dispatch(authLogOut(token));

  clearUser(dispatch);

  try {
    await client.logOut();
  } catch (e) {
    return true;
  }
};

/**
 * Persist user action
 *
 * Sets user data to local storage
 *
 * @param {Function} dispatch
 * @param {object} data
 */
const persistUser = (dispatch, data) => {
  const userUuid = prop('uuid', data);

  // If a user is 'pending' then throw a forbidden error and return
  if (propEq('status', UserStatusTypes.PENDING, data)) {
    return dispatch(errorAction(AUTH_REQUEST, { status: FORBIDDEN, unverified: true }));
  }

  setRememberedToken(userUuid);
  setRememberedUser(data);
  setRememberedCountry(propOr('GB', 'countryCode', data));

  dispatch(setToken(userUuid));
};

/**
 * Log in action
 *
 * @param {object} values
 * @returns {Function}
 */
export const logIn = values => async dispatch => {
  dispatch(authRequest(omit(['password'], values)));

  try {
    const {
      data: { data },
    } = await client.logIn({ data: { attributes: values } });

    persistUser(dispatch, data);
    dispatch(successAction(AUTH_REQUEST, { user: { ...data } }));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_REQUEST, e));
  }
};

/**
 * Reset password action
 *
 * @param {object} values
 * @returns {Function}
 */
export const resetPassword = values => async dispatch => {
  dispatch(authPasswordReset(values));

  try {
    await client.resetPassword({ data: { attributes: values } });
    dispatch(successAction(AUTH_PASSWORD_RESET, values));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SET_PASSWORD, e));
    throw e;
  }
};

/**
 * Set password action
 *
 * @param {object} values
 * @returns {Function}
 */
export const setPassword = values => async dispatch => {
  dispatch(authSetPasswordReset(omit(['values'], values)));

  try {
    await client.setPassword({ data: { attributes: values } });
    dispatch(successAction(AUTH_SET_PASSWORD, values));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_SET_PASSWORD, e));
  }
};

/**
 * Auth check action
 *
 * @param {object} params
 * @returns {Function}
 */
export const authCheck = params => async (dispatch, getState) => {
  dispatch(genericAction(AUTH_CHECK));

  const role = getCurrentUserType(getState());

  // Associations for auth call are different per role
  // TAs : assignedSalesRepresentatives
  // Everyone else : assignedTravelAgents
  const authParams = mergeDeepLeft(params, {
    associations: equals(AuthTypes.TA, role) ? 'assignedSalesRepresentatives' : 'assignedTravelAgents',
  });

  try {
    const {
      data: { data },
    } = await userClient.me(authParams);
    dispatch(successAction(AUTH_CHECK, { user: { ...data } }));
    persistUser(dispatch, data);
  } catch (e) {
    clearUser(dispatch);

    throw e;
  }
};

/**
 * Update password action
 *
 * @param {object} body
 * @returns {Function}
 */
export const updatePassword = body => async dispatch => {
  dispatch(genericAction(AUTH_PASSWORD_UPDATE, body));

  try {
    await client.updatePassword({ data: { attributes: body } });
    dispatch(successAction(AUTH_SET_PASSWORD, body));
    dispatch(enqueueNotification({ message: 'Password has been updated.', options: { variant: 'success' } }));
  } catch (e) {
    dispatch(errorFromResponse(AUTH_PASSWORD_UPDATE, e));
  }
};
