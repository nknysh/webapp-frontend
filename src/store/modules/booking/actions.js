import { __, pick, lensProp, lensPath, prop, set, view, map, mapObjIndexed, pipe, mergeDeepRight } from 'ramda';
import { eachDay, subDays } from 'date-fns';

import client from 'api/bookings';
import { formatDate } from 'utils';

import { successAction, errorFromResponse } from 'store/common';
import { enqueueNotification } from 'store/modules/ui/actions';
import { getHotelRoom } from 'store/modules/hotels/selectors';

import { getBookingData, getBookingByHotelId, getBookingRoomDatesById } from './selectors';

const accommodationProductsLens = lensProp('accommodationProducts');
const datesLens = lensProp('dates');
const ratesLens = lensPath(['dates', 'rates']);

export const BOOKING_UPDATE = 'BOOKING_UPDATE';
export const BOOKING_SUBMIT = 'BOOKING_SUBMIT';
export const BOOKING_REMOVE = 'BOOKING_REMOVE';

export const updateBookingAction = payload => ({
  type: BOOKING_UPDATE,
  payload,
});

export const submitBookingAction = payload => ({
  type: BOOKING_SUBMIT,
  payload,
});

export const removeBooking = payload => ({
  type: BOOKING_REMOVE,
  payload,
});

export const updateBooking = (id, payload) => (dispatch, getState) => {
  const state = getState();

  dispatch(updateBookingAction(payload));
  dispatch(successAction(BOOKING_UPDATE, mergeDeepRight(getBookingData(state), { [id]: payload })));
};

export const completeBooking = (id, payload) => async (dispatch, getState) => {
  dispatch(updateBooking(id, payload));

  const state = getState();
  const booking = getBookingByHotelId(state, id);

  const finaliseAccommodationProducts = (product, productId) => {
    const hotelRoom = getHotelRoom(state, id, productId);
    const { to, from } = getBookingRoomDatesById(state, id, productId);

    const rates = prop('rates', hotelRoom);

    const days = eachDay(from, subDays(to, 1));
    const ratesForDays = pipe(
      map(formatDate),
      pick(__, rates)
    )(days);

    return pipe(
      set(datesLens, { to, from }),
      set(ratesLens, ratesForDays)
    )(product);
  };

  const finalBooking = pipe(
    set(
      accommodationProductsLens,
      mapObjIndexed(finaliseAccommodationProducts, view(accommodationProductsLens, booking))
    )
  )(booking);

  dispatch(submitBookingAction(finalBooking));

  try {
    const {
      data: { data },
    } = await client.createBooking({ data: { attributes: { id, ...finalBooking } } });

    dispatch(updateBooking(id, { reservationId: prop('uuid', data) }));
    dispatch(successAction(BOOKING_SUBMIT, data));
  } catch (e) {
    dispatch(
      enqueueNotification({ message: 'There was a problem creating your booking.', options: { variant: 'error' } })
    );
    dispatch(errorFromResponse(BOOKING_SUBMIT, e));
    throw e;
  }
};
