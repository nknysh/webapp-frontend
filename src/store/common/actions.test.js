import { Status } from './status';
import { successAction, errorAction, extractData, fetchRelationships } from './actions';

describe('common actions', () => {
  it('successAction returns dynamic generated action object', () => {
    const action = successAction('FOO_ACTION', { some: 'data' });
    const expected = {
      type: `FOO_ACTION_${Status.SUCCESS}`,
      payload: { some: 'data' },
    };

    expect(action).toEqual(expected);
  });

  it('errorAction returns dynamic generated action object', () => {
    const action = errorAction('FOO_ACTION', { some: 'error' });
    const expected = {
      type: `FOO_ACTION_${Status.ERROR}`,
      payload: { some: 'error' },
    };

    expect(action).toEqual(expected);
  });

  describe('extractData', () => {
    it('returns array of unique relationship ids based on schema', () => {
      const entities = [
        {
          barId: 'foo',
        },
      ];

      const relationships = {
        bar: {
          path: ['barId'],
        },
      };

      expect(extractData(['path'], relationships, entities)).toMatchSnapshot();
    });
  });

  describe('fetchRelationships', () => {
    it('calls fetch actions for relationships if they do not exist', () => {
      const relationships = { foo: ['fooId1', 'fooId2'] };
      const actions = { foo: jest.fn() };
      const getState = () => ({
        foo: {
          data: {
            fooId2: {
              title: 'foo 1',
            },
          },
        },
      });
      const dispatch = jest.fn();

      fetchRelationships(relationships, actions, getState, dispatch);

      expect(actions.foo).toHaveBeenCalledTimes(1);
      expect(actions.foo).toHaveBeenCalledWith('fooId1');
      expect(dispatch).toHaveBeenCalledTimes(1);
    });
  });
});
