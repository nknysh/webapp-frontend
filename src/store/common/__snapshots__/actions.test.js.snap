// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common actions extractData returns array of unique relationship ids based on schema 1`] = `
Object {
  "bar": Array [
    "foo",
  ],
}
`;
